feat(ui): bulk remove vulnerabilities

0b3ud opened the pull request:
* Add Feature #43
#### Reading this issue I was able to split this feature in two tasks :

- [x] Selecting all vulnerabilities in vulnerability page
- [x] Bulk deleting selected vulnerabilities 

####  Selecting all vulnerabilities in vulnerability page : 

1. Edited template in web/templates/base/_items/vulnerability_tab_content.html : 

- > Added input called `select_all_checkbox`
![image](https://github.com/user-attachments/assets/93346b7c-b3de-4229-ac6a-9f0eca96cf0e)



<br>

2. Bulk deleting selected vulnerabilities : 

- > Added a button for bulk deleting in template  <br> File : web/templates/base/_items/vulnerability_tab_content.html 
![image](https://github.com/user-attachments/assets/9f6908e8-2f4f-432a-9566-21f0444ea7b9)


#### Results : 

- I've tested both function using the results of a scan I did on a metasploitable machine :100: 
- Selecting all seems to work :+1: 
- Bulk deletion seems to work :+1: 

I would appreciate it if you can do some tests on your side and notify me if there are any bugs

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

Implement bulk operations for vulnerabilities, including selecting all, resolving, and deleting multiple vulnerabilities at once. Enhance the user interface to support these operations and improve the dynamic status update of vulnerabilities.

New Features:
- Introduce a feature to select all vulnerabilities on the vulnerability page, allowing users to manage multiple vulnerabilities at once.
- Add functionality for bulk deleting selected vulnerabilities, streamlining the process of removing multiple vulnerabilities simultaneously.

Enhancements:
- Improve the user interface by adding buttons for bulk operations such as opening, resolving, and deleting selected vulnerabilities.
- Enhance the vulnerability status change function to update the UI dynamically, reflecting the current status of vulnerabilities.

<!-- Generated by sourcery-ai[bot]: end summary -->

AnonymousWP commented:
Any reason why some of the content of your PR description is quoted?

0b3ud commented:
Nope just wanted it to be easier to read 
It turned out to be confusing my bad

psyray commented on web/startScan/templates/startScan/vulnerabilities.html:
```diff
@@ -79,10 +79,10 @@
 				"render": function ( data, type, row ) {
 					if(row['open_status']){
-						return `<div class="form-check mb-2 form-check-primary"><input type="checkbox" name="targets_checkbox['+ e + ']" class="float-start form-check-input" onchange="vuln_status_change(this, `+data+`);">\n<span class="new-control-indicator"></span><span style="visibility:hidden">c</span></div>`;
+						return `<div class="form-check mb-2 form-check-primary"><input type="checkbox" name="targets_checkbox['+ e + ']" class="float-start form-check-input vulnerability_checkbox" onchange="vuln_status_change(this, `+data+`);">\n<span class="new-control-indicator"></span><span style="visibility:hidden">c</span></div>`;
```
You must move the onchange event to a onclick event on the Status badge  on the end of the line
![image](https://github.com/user-attachments/assets/b166b86f-9ad8-4853-bbab-d23da1f00c90)

Because when item is clicked, status is changed
![image](https://github.com/user-attachments/assets/2855f619-a3cd-47c6-a849-12a612208e64)

psyray replied:
Resolved, perfect :+1:

psyray commented on web/templates/base/_items/vulnerability_tab_content.html:
```diff
@@ -21,6 +21,17 @@
+							</div>
+						</div>
+						<div class="btn-group mb-2 float-end dropstart">
+							<button id="bulk_delete_vulnerabilities" class="btn btn-soft-primary ms-1" aria-haspopup="true" aria-expanded="false" data-toggle="tooltip" data-placement="left" title="Bulk Delete Vulnerabilities">
```
Could you put the trash button in red

psyray replied:
Resolved :+1:

psyray commented on web/templates/base/_items/vulnerability_tab_content.html:
```diff
@@ -21,6 +21,17 @@
 			<div class="card-body">
 				<div class="row">
 					<div class="col-12">
```
Could you add before the delete button, a green status button to permit mass update of the vuln status

0b3ud replied:
Sure on it

psyray replied:
Resolved

0b3ud commented:
## Sept 4th 2024 Update :
### This PR  will add the following features :
- [ ] Selecting all vulnerabilities in vulnerability page
- [ ] Bulk deleting selected vulnerabilities 
- [ ] Bulk changing vulnerabilities status
- [ ] Status Badge is clickable 

![Screenshot from 2024-09-04 12-05-23](https://github.com/user-attachments/assets/0b16c6e3-d814-4bcc-ad45-dd50fe7b3135)

- I have changed the color of the trash button and put it in red
- Collaborated with @psyray to make the status badges clickable to change vulnerability status
- Added buttons resolve all and open all to change status of selected vulnerabilities
 

On my behalf I have tested it, and it seems to work :100: 
Let me know if you identify any bugs :1st_place_medal:

psyray reviewed (changes requested):
@0b3ud Thanks for this.

To be consistent with other tables could you move the mass actions buttons to the left side like in the subdomain table
![image](https://github.com/user-attachments/assets/136e8424-6478-4a3c-afcc-1e05bf2c2030)

And move the select all checkbox at the top of the table
![image](https://github.com/user-attachments/assets/a27fc480-fbfc-41bb-9c70-ae415fd4a197)

0b3ud commented:
Hey
Thanks for the review
Sure on it


> @0b3ud Thanks for this.
> 
> Could you use the Blue color for the open state, red is already used for vuln severity and it could be confusing
> 
> To be consistent with other tables could you move the mass actions buttons to the left side like in the subdomain table ![image](https://private-user-images.githubusercontent.com/1230954/365181244-136e8424-6478-4a3c-afcc-1e05bf2c2030.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MzkxODcsIm5iZiI6MTcyNTYzODg4NywicGF0aCI6Ii8xMjMwOTU0LzM2NTE4MTI0NC0xMzZlODQyNC02NDc4LTRhM2MtYWZjYy0xZTA1YmYyYzIwMzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTYwODA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjlhZjNiNmZmMDkwNTBkNTc4YzYxN2UxNDlkMTRlMDdjYTJkNzE2MDgxZTgwYTA2MTI4Yjg5ZTAyYmU5MzY5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.32ftDxF0dRwQlzdb6V3WjdmasTP1wn-OqJ7065gUSDc)
> 
> And move the select all checkbox at the top of the table ![image](https://private-user-images.githubusercontent.com/1230954/365181472-a27fc480-fbfc-41bb-9c70-ae415fd4a197.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MzkxODcsIm5iZiI6MTcyNTYzODg4NywicGF0aCI6Ii8xMjMwOTU0LzM2NTE4MTQ3Mi1hMjdmYzQ4MC1mYmZjLTQxYmItOWM3MC1hZTQxNWZkNGExOTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTYwODA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWJkZGQxYjljNTBhNTQzZmE0M2U5ODg5ODNiMmVmODhkYjNmNWYxYjkyYzkwMzA2MzUyZDBlZWE2OTFkNjM3MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.iquEGxVyK0AEuc-sX8kLsdaqsWGneiTt_pU61kRUg90)

0b3ud commented:
## Sept 6th Update :

### As requested I have changed the following : 

- [ ] Blue color for the open state
- [ ] The mass actions buttons moved to the left side like in the subdomain table 
- [ ] Select all button is now in the table header
- [ ] Changed some text and Added some icons :) 

![image](https://github.com/user-attachments/assets/6f38dca0-85e3-45b8-8201-4188bf442e15)

AnonymousWP commented:
@0b3ud Can you check the checkboxes where applicable? e.g. beginning of this PR and your latest post. Not checking them indicates it's not done.

0b3ud commented:
> @0b3ud Can you check the checkboxes where applicable? e.g. beginning of this PR and your latest post. Not checking them indicates it's not done.

Done :)

psyray commented:
Don't have tested yet.
Tomorrow

psyray commented:
@sourcery-ai review

AnonymousWP commented:
Please finish this up so I can test this (after sourcery's suggestions).

0b3ud commented:
> Please finish this up so I can test this (after sourcery's suggestions).

Done you can test it now :)

psyray commented:
> > Please finish this up so I can test this (after sourcery's suggestions).
> 
> Done you can test it now :)

Near perfect, just one thing, you need to disable mass change buttons if no checkbox is checked.
And if it's checked you display the number of checked vulns
Like in the subdomains list

Disabled
![image](https://github.com/user-attachments/assets/583531ef-46e1-4bd6-b0a7-a36a12a205ad)

Enabled
![image](https://github.com/user-attachments/assets/c8499a81-5e57-40a9-95ff-4aee79a394fc)

```js
function onchange(event) {
  toggleMultipleSubdomainButton()
}
function toggleMultipleSubdomainButton() {
  var checked_count = checkedCount();
  if (checked_count > 0) {
    $("[data-button=subdomain_btns]").removeClass("disabled");
    $('#subdomain_selected_count').text(checked_count + ' Subdomains Selected x');
  } else {
    $("[data-button=subdomain_btns]").addClass("disabled");
    $('#subdomain_selected_count').empty();
  }
}
```

0b3ud commented:
> > > Please finish this up so I can test this (after sourcery's suggestions).
> > 
> > 
> > Done you can test it now :)
> 
> Near perfect, just one thing, you need to disable mass change buttons if no checkbox is checked. And if it's checked you display the number of checked vulns Like in the subdomains list
> 
> Disabled ![image](https://private-user-images.githubusercontent.com/1230954/368715802-583531ef-46e1-4bd6-b0a7-a36a12a205ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY4MTg1MzcsIm5iZiI6MTcyNjgxODIzNywicGF0aCI6Ii8xMjMwOTU0LzM2ODcxNTgwMi01ODM1MzFlZi00NmUxLTRiZDYtYjBhNy1hMzZhMTJhMjA1YWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjBUMDc0MzU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MThmNTU2YTBmN2U5MDM1NGQ5N2M4ODQ2NDdkZDg2MTIzYzZhMTY0NjQ3NDhhNWFmMDMwYjkwOGVlNDQ5ZGY0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WmJJ5aqqOnZY9hRErJYzoQ15Mk2Bp0bbBcYynTxpMg8)
> 
> Enabled ![image](https://private-user-images.githubusercontent.com/1230954/368715960-c8499a81-5e57-40a9-95ff-4aee79a394fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY4MTg1MzcsIm5iZiI6MTcyNjgxODIzNywicGF0aCI6Ii8xMjMwOTU0LzM2ODcxNTk2MC1jODQ5OWE4MS01ZTU3LTQwYTktOTVmZi00YWVlNzlhMzk0ZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjBUMDc0MzU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjQyOTJhNzNlYWRlOWZjMDI4N2NlNzA2NjhlY2EzYTZkMjVhZTRjZmEwZTQzOTBmZDdlMjk0ZTNiYmVhMmYwZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UIBuVPhOWrSjJvuETPeSuwgqmjkAxXiG2PBKzz-AAPU)
> 
> ```js
> function onchange(event) {
>   toggleMultipleSubdomainButton()
> }
> function toggleMultipleSubdomainButton() {
>   var checked_count = checkedCount();
>   if (checked_count > 0) {
>     $("[data-button=subdomain_btns]").removeClass("disabled");
>     $('#subdomain_selected_count').text(checked_count + ' Subdomains Selected x');
>   } else {
>     $("[data-button=subdomain_btns]").addClass("disabled");
>     $('#subdomain_selected_count').empty();
>   }
> }
> ```


Hey @psyray 
Would you able to check 
I have added the feature as demanded :)

0b3ud commented on web/templates/base/_items/vulnerability_tab_content.html:
```diff
@@ -90,11 +108,16 @@
 							</div>
 						</div>
 					</div>
+					<div class="ms-2 mt-2" id="">
```
fixed
